tests/31sync/02sync.pl

```diff
@@ -1,3 +1,222 @@
+use Future::Utils qw( repeat );
+
+push our @EXPORT, qw(
+   matrix_do_and_wait_for_sync
+   matrix_send_room_text_message_and_wait_for_sync
+   matrix_send_room_message_and_wait_for_sync
+   matrix_create_room_and_wait_for_sync
+   matrix_join_room_and_wait_for_sync
+   matrix_leave_room_and_wait_for_sync
+   matrix_invite_user_to_room_and_wait_for_sync
+   matrix_put_room_state_and_wait_for_sync
+   matrix_advance_room_receipt_and_wait_for_sync
+   sync_timeline_contains
+);
+
+sub matrix_do_and_wait_for_sync
+{
+   my ( $user, %params ) = @_;
+
+   my $check = delete $params{check} or die "Must supply a 'check' param";
+   my $do = delete $params{do} or die "Must supply a 'do' param";
+   $params{timeout} = $params{timeout} // 1000;
+
+   my $next_batch;
+
+   matrix_sync( $user,
+      filter            => '{"room":{"rooms":[]},"account_data":{"types":[]},"presence":{"types":[]}}',
+      update_next_batch => 0,
+      set_presence      => "offline",
+   )->then( sub {
+      my ( $body ) = @_;
+
+      $next_batch = $body->{next_batch};
+
+      $do->();
+   })->then( sub {
+      my ( $action_result ) = @_;
+
+      my $finished = repeat {
+            matrix_sync( $user,
+               %params,
+               since             => $next_batch,
+               update_next_batch => 0,
+               set_presence      => "offline",
+            )->then( sub {
+               my ( $body ) = @_;
+
+               $next_batch = $body->{next_batch};
+
+               Future->done( $check->( $body, $action_result ) );
+            });
+         }
+         until => sub {
+            $_[0]->failure or $_[0]->get
+         };
+
+      $finished->then( sub { Future->done( $action_result ); } );
+   });
+}
+
+sub sync_room_contains
+{
+   my ( $sync_body, $room_id, $section, $check ) = @_;
+
+   my $room =  $sync_body->{rooms}{join}{$room_id};
+
+   foreach my $event ( @{ $room->{$section}{events} } ) {
+      if ( $check->( $event ) ) {
+         return 1;
+      }
+   }
+
+   return 0;
```

Comment:
This whole chunk could just be

    return any { $check->( $_ ) } @{ $room->{$section}{events} };